zebra/zserv.c

```diff
@@ -314,17 +320,17 @@ static void zserv_read(struct event *thread)
 	size_t already;
 	struct stream_fifo *cache;
 	uint32_t p2p_orig;
-
 	uint32_t p2p;
 	struct zmsghdr hdr;
+	size_t client_ibuf_fifo_cnt = stream_fifo_count_safe(client->ibuf_fifo);
 
 	p2p_orig = atomic_load_explicit(&zrouter.packets_to_process,
 					memory_order_relaxed);
 	cache = stream_fifo_new();
 	p2p = p2p_orig;
 	sock = EVENT_FD(thread);
 
-	while (p2p) {
+	while (p2p - client_ibuf_fifo_cnt) {
```

Comment:
Don't love treating the arithmetic result as a boolean - I'd rather see this a little clearer. Could we adjust `p2p` initially, then just decrement and test it as the code does now?
Also a little concerned about the implicit assumption that we'll never see a condition where the fifo is too large - after a change to the `packets_to_process` for example. I'd rather see that protected against: if the fifo is "full", `p2p` is just zero and we return.

Reply:
Will make this change.